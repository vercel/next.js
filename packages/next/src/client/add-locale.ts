import type { addLocale as Fn } from '../shared/lib/router/utils/add-locale'
import { normalizePathTrailingSlash } from './normalize-trailing-slash'

export const addLocale: typeof Fn = (path, ...args) => {
  if (process.env.__NEXT_I18N_SUPPORT) {
    return normalizePathTrailingSlash(
      (
        require('../shared/lib/router/utils/add-locale') as typeof import('../shared/lib/router/utils/add-locale')
      ).addLocale(path, ...args)
    )
  }
  return path
}
