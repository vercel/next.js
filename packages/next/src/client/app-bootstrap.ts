/**
 * Before starting the Next.js runtime and requiring any module, we need to make
 * sure the following scripts are executed in the correct order:
 * - Polyfills
 * - next/script with `beforeInteractive` strategy
 */

const version = process.env.__NEXT_VERSION

window.next = {
  version,
  appDir: true,
}

function loadScriptsInSequence(
  scripts: [src: string, props: { [prop: string]: any }][],
  hydrate: () => void
) {
  if (!scripts || !scripts.length) {
    return hydrate()
  }

  return scripts
    .reduce((promise, [src, props]) => {
      return promise.then(() => {
        return new Promise<void>((resolve, reject) => {
          const el = document.createElement('script')

          if (props) {
            for (const key in props) {
              if (key !== 'children') {
                el.setAttribute(key, props[key])
              }
            }
          }

          if (src) {
            el.src = src
            el.onload = () => resolve()
            el.onerror = reject
          } else if (props) {
            el.innerHTML = props.children
            setTimeout(resolve)
          }

          document.head.appendChild(el)
        })
      })
    }, Promise.resolve())
    .catch((err: Error) => {
      console.error(err)
      // Still try to hydrate even if there's an error.
    })
    .then(() => {
      hydrate()
    })
}

export function appBootstrap(hydrate: () => void) {
  loadScriptsInSequence((self as any).__next_s, () => {
    // If the static shell is being debugged, skip hydration if the
    // `__nextppronly` query is present. This is only enabled when the
    // environment variable `__NEXT_EXPERIMENTAL_STATIC_SHELL_DEBUGGING` is
    // set to `1`. Otherwise the following is optimized out.
    if (process.env.__NEXT_EXPERIMENTAL_STATIC_SHELL_DEBUGGING === '1') {
      const search = new URLSearchParams(window.location.search)
      if (
        search.get('__nextppronly') === 'fallback' ||
        search.get('__nextppronly') === '1'
      ) {
        console.warn(
          `Skipping hydration due to __nextppronly=${search.get('__nextppronly')}`
        )
        return
      }
    }

    hydrate()
  })
}
