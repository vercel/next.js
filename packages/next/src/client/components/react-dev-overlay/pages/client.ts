import * as Bus from './bus'
import { parseStack } from '../internal/helpers/parse-stack'
import { parseComponentStack } from '../internal/helpers/parse-component-stack'
import {
  hydrationErrorState,
  storeHydrationErrorStateFromConsoleArgs,
} from '../internal/helpers/hydration-error-info'
import {
  ACTION_BEFORE_REFRESH,
  ACTION_BUILD_ERROR,
  ACTION_BUILD_OK,
  ACTION_REFRESH,
  ACTION_UNHANDLED_ERROR,
  ACTION_UNHANDLED_REJECTION,
  ACTION_VERSION_INFO,
} from '../shared'
import type { VersionInfo } from '../../../../server/dev/parse-version-info'
import { attachHydrationErrorState } from '../internal/helpers/attach-hydration-error-state'

let isRegistered = false
let stackTraceLimit: number | undefined = undefined

function handleError(error: unknown) {
  if (!error || !(error instanceof Error) || typeof error.stack !== 'string') {
    // A non-error was thrown, we don't have anything to show. :-(
    return
  }

  attachHydrationErrorState(error)

  const componentStackTrace =
    (error as any)._componentStack || hydrationErrorState.componentStack
  const componentStackFrames =
    typeof componentStackTrace === 'string'
      ? parseComponentStack(componentStackTrace)
      : undefined

  // Skip ModuleBuildError and ModuleNotFoundError, as it will be sent through onBuildError callback.
  // This is to avoid same error as different type showing up on client to cause flashing.
  if (
    error.name !== 'ModuleBuildError' &&
    error.name !== 'ModuleNotFoundError'
  ) {
    Bus.emit({
      type: ACTION_UNHANDLED_ERROR,
      reason: error,
      frames: parseStack(error.stack),
      componentStackFrames,
    })
  }
}

let origConsoleError = console.error
function nextJsHandleConsoleError(...args: any[]) {
  // See https://github.com/facebook/react/blob/d50323eb845c5fde0d720cae888bf35dedd05506/packages/react-reconciler/src/ReactFiberErrorLogger.js#L78
  const error = process.env.NODE_ENV !== 'production' ? args[1] : args[0]
  storeHydrationErrorStateFromConsoleArgs(...args)
  handleError(error)
  origConsoleError.apply(window.console, args)
}

function onUnhandledError(event: ErrorEvent) {
  const error = event?.error
  handleError(error)
}

function onUnhandledRejection(ev: PromiseRejectionEvent) {
  const reason = ev?.reason
  if (
    !reason ||
    !(reason instanceof Error) ||
    typeof reason.stack !== 'string'
  ) {
    // A non-error was thrown, we don't have anything to show. :-(
    return
  }

  const e = reason
  Bus.emit({
    type: ACTION_UNHANDLED_REJECTION,
    reason: reason,
    frames: parseStack(e.stack!),
  })
}

export function register() {
  if (isRegistered) {
    return
  }
  isRegistered = true

  try {
    const limit = Error.stackTraceLimit
    Error.stackTraceLimit = 50
    stackTraceLimit = limit
  } catch {}

  window.addEventListener('error', onUnhandledError)
  window.addEventListener('unhandledrejection', onUnhandledRejection)
  window.console.error = nextJsHandleConsoleError
}

export function unregister() {
  if (!isRegistered) {
    return
  }
  isRegistered = false

  if (stackTraceLimit !== undefined) {
    try {
      Error.stackTraceLimit = stackTraceLimit
    } catch {}
    stackTraceLimit = undefined
  }

  window.removeEventListener('error', onUnhandledError)
  window.removeEventListener('unhandledrejection', onUnhandledRejection)
  window.console.error = origConsoleError
}

export function onBuildOk() {
  Bus.emit({ type: ACTION_BUILD_OK })
}

export function onBuildError(message: string) {
  Bus.emit({ type: ACTION_BUILD_ERROR, message })
}

export function onRefresh() {
  Bus.emit({ type: ACTION_REFRESH })
}

export function onBeforeRefresh() {
  Bus.emit({ type: ACTION_BEFORE_REFRESH })
}

export function onVersionInfo(versionInfo: VersionInfo) {
  Bus.emit({ type: ACTION_VERSION_INFO, versionInfo })
}

export { getErrorByType } from '../internal/helpers/get-error-by-type'
export { getServerError } from '../internal/helpers/node-stack-frames'
export { default as ReactDevOverlay } from './ReactDevOverlay'
