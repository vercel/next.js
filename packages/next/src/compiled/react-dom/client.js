'use strict';

function checkDCE() {
  /* global __REACT_DEVTOOLS_GLOBAL_HOOK__ */
  if (
    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ === 'undefined' ||
    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.checkDCE !== 'function'
  ) {
    return;
  }
  if (process.env.NODE_ENV !== 'production') {
    // This branch is unreachable because this function is only called
    // in production, but the condition is true only in development.
    // Therefore if the branch is still here, dead code elimination wasn't
    // properly applied.
    // Don't change the message. React DevTools relies on it. Also make sure
    // this message doesn't occur elsewhere in this function, or it will cause
    // a false positive.
    throw new Error('^_^');
  }
  try {
    // Verify that the code above has been dead code eliminated (DCE'd).
    __REACT_DEVTOOLS_GLOBAL_HOOK__.checkDCE(checkDCE);
  } catch (err) {
    // DevTools shouldn't crash React, no matter what.
    // We should still report in case we break this code.
    console.error(err);
  }
}

if (process.env.NODE_ENV === 'production') {
  // DCE check should happen before ReactDOM bundle executes so that
  // DevTools can report bad minification during injection.
  checkDCE();
  module.exports = require('./cjs/react-dom-client.production.js');
} else {
  module.exports = require('./cjs/react-dom-client.development.js');
}
