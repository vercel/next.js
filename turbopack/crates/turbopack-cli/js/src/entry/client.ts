import { connect } from "@vercel/turbopack-ecmascript-runtime/browser/dev/hmr-client/hmr-client";
import {
  connectHMR,
  addMessageListener,
  sendMessage,
} from "@vercel/turbopack-ecmascript-runtime/browser/dev/hmr-client/websocket";

export function initializeHMR(options: { assetPrefix: string }) {
  connect({
    addMessageListener,
    sendMessage,
    onUpdateError: console.error,
  });
  connectHMR({
    assetPrefix: options.assetPrefix,
    log: true,
    path: "/turbopack-hmr",
  });
}
