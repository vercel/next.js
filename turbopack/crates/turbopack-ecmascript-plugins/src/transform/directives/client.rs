use anyhow::Result;
use async_trait::async_trait;
use swc_core::ecma::{ast::Program, transforms::base::resolver, visit::VisitMutWith};
use turbo_rcstr::RcStr;
use turbo_tasks::Vc;
use turbopack_ecmascript::{CustomTransformer, TransformContext};

use super::{is_client_module, server_to_client_proxy::create_proxy_module};

#[derive(Debug)]
pub struct ClientDirectiveTransformer {
    transition_name: Vc<RcStr>,
}

impl ClientDirectiveTransformer {
    pub fn new(transition_name: Vc<RcStr>) -> Self {
        Self { transition_name }
    }
}

#[async_trait]
impl CustomTransformer for ClientDirectiveTransformer {
    #[tracing::instrument(level = tracing::Level::TRACE, name = "client_directive", skip_all)]
    async fn transform(&self, program: &mut Program, ctx: &TransformContext<'_>) -> Result<()> {
        if is_client_module(program) {
            let transition_name = &*self.transition_name.await?;
            *program = create_proxy_module(transition_name, &format!("./{}", ctx.file_name_str));
            program.visit_mut_with(&mut resolver(
                ctx.unresolved_mark,
                ctx.top_level_mark,
                false,
            ));
        }

        Ok(())
    }
}
