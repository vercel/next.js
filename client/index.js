/* global __NEXT_DATA__ */
import { createElement, startTransition } from "react";
import mitt from "mitt";
import { waitForPage } from "../lib/page-loader";
import router, { createRouter } from "./router";
import { getURL } from "../lib/url";
import { createRoot, hydrateRoot } from "react-dom/client";

const { props, err, pathname, query } = __NEXT_DATA__;

__webpack_public_path__ = __NEXT_DATA__.publicPath; // eslint-disable-line

const asPath = getURL();

const appContainer = document.getElementById("__next");

let lastAppProps;
let ErrorComponent;
let reactRoot = null;

export const emitter = mitt();

let Enhancer;
export function setEnhancer(_enhancer) {
  Enhancer = _enhancer;
}

export default () => {
  return Promise.all([
    waitForPage("/_error"),
    waitForPage(pathname).catch(console.error),
  ]).then(([_ErrorComponent, Component]) => {
    console.log("clinent/index.js error component");
    ErrorComponent = _ErrorComponent;
    Component = Component || ErrorComponent;

    createRouter(
      pathname,
      query,
      asPath,
      {
        Component,
        err,
      },
      render
    );

    render({ Component, props, err });

    return emitter;
  });
};

export function render({ Component, props, err }) {
  // There are some errors we should ignore.
  // Next.js rendering logic knows how to handle them.
  // These are specially 404 errors
  if (err && !err.ignore) {
    console.log("rendering erorr client/index.js", err);
    return renderError(err);
  }

  let loadProps;
  if (
    !props &&
    Component &&
    Component !== ErrorComponent &&
    lastAppProps.Component === ErrorComponent
  ) {
    // fetch props if ErrorComponent was replaced with a page component by HMR
    const { pathname, query, asPath } = router.url;
    loadProps = Component.getInitialProps({ err, pathname, query, asPath });
  } else {
    loadProps = Promise.resolve(props);
  }

  return loadProps
    .then((props) => {
      Component = Component || lastAppProps.Component;
      props = props || lastAppProps.props;

      const appProps = { Component, props, err, router };
      // lastAppProps has to be set before ReactDom.render to account for ReactDom throwing an error.
      lastAppProps = appProps;

      emitter.emit("before-reactdom-render", {
        Component,
        ErrorComponent,
        appProps,
      });
      renderReactElement(
        createElement(Component, { url: router.url, ...props }),
        appContainer
      );
      emitter.emit("after-reactdom-render", {
        Component,
        ErrorComponent,
        appProps,
      });
    })
    .catch((err) => {
      console.log("err in loadProps", err);
      if (err.abort) return;
      return renderError(err);
    });
}

// This method handles all runtime and debug errors.
// 404 and 500 errors are special kind of errors
// and they are still handle via the main render method.
export function renderError(error) {
  // We need to unmount the current app component because it's
  // in the inconsistant state.
  // Otherwise, we need to face issues when the issue is fixed and
  // it's get notified via HMR
  // reactRoot.unmount();
  console.log("error in renderError function ", error);

  console.error(error);

  return ErrorComponent.getInitialProps({
    err: error,
    pathname,
    query,
    asPath,
  }).then((props) => {
    const appProps = { Component: ErrorComponent, props, err: error, router };
    console.log("emiting before-reactdom-render", err, pathname, query, asPath);
    emitter.emit("before-reactdom-render", { ErrorComponent, appProps });
    renderReactElement(createElement(ErrorComponent, props), appContainer);
    emitter.emit("after-reactdom-render", { ErrorComponent, appProps });
  });
}

// dom ell is the app container
function renderReactElement(reactEl, domEl) {
  // Wrap page in app-level enhancer, if defined
  reactEl = Enhancer ? createElement(Enhancer, null, reactEl) : reactEl;

  // it looks like root needs to be similar to react root so it's checking it
  // which is basically what isInitialRender is doing so maybe i can try getting rid of that so we dont have to save the root

  // do we need a root to render ?
  // how to render an error for the page?

  console.log("reactEl is ", reactEl);

  if (!reactRoot) {
    // todo: do we need to save the root?
    reactRoot = hydrateRoot(domEl, reactEl);
  } else {
    // start transition schedules an update to react tree
    startTransition(() => {
      reactRoot.render(reactEl);
    });
  }
}

// If the root has not been defined yet and that means it's initial render then we need to hydrate

// if the root has already been defined that means we should handle updating and use the starttransition function and update hthe tree/ render it
