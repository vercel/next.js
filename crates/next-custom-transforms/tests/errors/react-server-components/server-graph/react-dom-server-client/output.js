// This is a comment.
'use strict';
/**
 * This is a comment.
 */ import 'react-dom/server';
import 'react-dom/client';
export default function() {
    return null;
}
