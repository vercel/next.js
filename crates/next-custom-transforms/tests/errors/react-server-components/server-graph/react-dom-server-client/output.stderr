  x You're importing a component that imports react-dom/server. To fix it, render or return the content directly as a Server Component instead for perf and security.
  | Learn more: https://nextjs.org/docs/app/building-your-application/rendering
   ,-[input.js:9:1]
 8 | 
 9 | import 'react-dom/server'
   : ^^^^^^^^^^^^^^^^^^^^^^^^^
   `----
  x You're importing a component that imports react-dom/client. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.
  | \nLearn more: https://nextjs.org/docs/app/building-your-application/rendering\n\n
    ,-[input.js:11:1]
 10 | 
 11 | import 'react-dom/client'
    : ^^^^^^^^^^^^^^^^^^^^^^^^^
    `----
